dask/array/slicing.py

```diff
@@ -813,3 +826,57 @@ def check_index(ind, dimension):
     elif ind < -dimension:
         msg = "Negative index is not greater than negative dimension %d <= -%d"
         raise IndexError(msg % (ind, dimension))
+
+
+def slice_with_dask_array(x, index):
+    from .core import Array, atop, elemwise
+
+    out_index = [slice(None)
+                 if isinstance(ind, Array) and ind.dtype == bool
+                 else ind
+                 for ind in index]
+
+    if len(index) == 1 and index[0].ndim == x.ndim:
+        y = elemwise(getitem, x, *index, dtype=x.dtype)
+        name = 'getitem-' + tokenize(x, index)
+        dsk = {(name, i): k for i, k in enumerate(core.flatten(y._keys()))}
+        chunks = ((np.nan,) * y.npartitions,)
+        return (Array(sharedict.merge(y.dask, (name, dsk)), name, chunks, x.dtype),
+                out_index)
+
+    if any(isinstance(ind, Array) and ind.dtype == bool and ind.ndim != 1
+            for ind in index):
+        raise NotImplementedError("Slicing with dask.array only permitted for "
```

Comment:
I found this error message a bit confusing, I think the "only permitted for dimension 1 and dimension n". Maybe something like "only permitted when the indexer has 1 dimension, or the dimensions of the indexer match the dimensions of the array being sliced."? It's a bit wordier.

Reply:
Fixed